### DISCUSSION TITLE
How to validate access token? {:error,  {:no_matching_key_with_kid, ...}}

### DISCUSSION THREAD

**Discussion by 1player** [erlef/oidcc#404] [Q&A] [answered]
I want to validate an access token I receive in a request, but it fails with `{:error,
 {:no_matching_key_with_kid, ...}}` even though I can see the kid in the provider configuration.

Example:

```elixir
{:ok, client_context} = Oidcc.ClientContext.from_configuration_worker(Gateway.OidccConfigProvider, "client-id", "client-secret") 
Oidcc.Token.validate_jwt(access_token, client_context, %{
  signing_algs:
    client_context.provider_configuration.token_endpoint_auth_signing_alg_values_supported
})
# Fails with {:error,
 {:no_matching_key_with_kid, "zgeOJ-OVB1Ewd6M5nSDjftW7uXjJkxq17dWaPo7z420"}}
```

Yet when I inspect the JWK set in the configuration with `Oidcc.ProviderConfiguration.Worker.get_jwks(Gateway.OidccConfigProvider)`, that kid is present in the list:

```elixir
#JOSE.JWK<
  keys: {:jose_jwk_set,
   [
     {:jose_jwk, :undefined,
      {:jose_jwk_kty_rsa,
       {:RSAPublicKey, snip, 65537}},
      %{
        "alg" => "RS256",
        "kid" => "zgeOJ-OVB1Ewd6M5nSDjftW7uXjJkxq17dWaPo7z420",
        "use" => "sig",
        "x5c" => ["snip"],
        "x5t" => "91MQNZV0Z1N-6TyBShUyUl1rrzo",
        "x5t#S256" => "Cr5db1sZskJFmCiQdcwsgU4CFZ5v6ZY5ueF2FfvEv10"
      }}
   ]},

```

Why is it failing, what am I missing?

Basically I'd like to replicate the flow described at https://stackoverflow.com/a/71183948

**Comment by maennchen** [comment]
Hm, the verification code is a bit unclear here since the errors aren't handled nicely by our JWT library.

Can you add some debug statements to the following lines so that we can see with what reason it is failing?

https://github.com/erlef/oidcc/blob/ef11cc3187e46503034d4518fe84ec0df9c4495d/src/oidcc_jwt_util.erl#L355

Something like this:

```erlang
error:Reason ->
  erlang:display(Reason),
  {error, {no_matching_key_with_kid, Kid}}
```

**Reply by 1player** [reply]
This is actually where it fails: https://github.com/erlef/oidcc/blob/ef11cc3187e46503034d4518fe84ec0df9c4495d/src/oidcc_jwt_util.erl#L97 

What other info can I provide to help?

Also, I swear I ran `mix test` on this project, and the first time it failed a doctest with exactly that same error, which I am unable to reproduce any more.

**Reply by 1player** [reply]
I dumped the arguments of that validate_signature call:

```
Token = <<"eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJ6Z2VPSi1PVkIxRXdkNk01blNEamZ0Vzd1WGpKa3hxMTdkV2FQbzd6NDIwIn0.eyJleHAiOjE3MzI2MjMyNzksImlhdCI6MTczMjYyMjk3OSwiYXV0aF90aW1lIjoxNzMyNjIyOTc5LCJqdGkiOiJhOGVmZTJiOS05MDUzLTRhOTctODkzZC1mYjkzODYxYTk2NzIiLCJpc3MiOiJodHRwczovL2lhbS5pdnYuZGVzcC5zcGFjZS9yZWFsbXMvZGVzcCIsInN1YiI6IjlkYWM1YzE4LWFmZGUtNGViYi1hN2VkLTNjM2ZhZmRmZGFiMyIsInR5cCI6IkJlYXJlciIsImF6cCI6InMzZ2F0ZXdheS1wdWJsaWMiLCJzZXNzaW9uX3N0YXRlIjoiZWIxNDE1MGItYTE4ZC00NDkwLWJjNjgtZGJlOThjMmU2OTUyIiwic2NvcGUiOiJvcGVuaWQiLCJzaWQiOiJlYjE0MTUwYi1hMThkLTQ0OTAtYmM2OC1kYmU5OGMyZTY5NTIiLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJzdGVwaGFuZS50In0.fQzFjZst2xE0mY35lEMCJ3D3JpcdFvNAdI6J9HVfr-6bHF9WFr5w46tcoY5u-gCTw8e1gG03JF3qUXNbcRHpxP2oHui5FBnmo9AeWyTD342M6lGLxOm3EiujqDILrrv8zC-JITYs2PuDPt4Qlweir2Rv-H5lwaqxA66SeCtRa_Pd1RWbyxrvCm9LfFMq2QjrX61-tbMOrZDCQSADtciO9d3cAlEyVpLIB7RhAcMg5sv6iC3DBOH54NuZZSpFZoMLlKeXYLqF0uL1fDYI1oSgAECuyYPJJykDbNPqtReckafV1Dgks1qrPfdYJc7wf49lKceGiTwkQoALv10Xhy9Mhw">>

Jwks = {jose_jwk,undefined,{jose_jwk_kty_rsa,{'RSAPublicKey',21966476291801841635828048148208692901386442529292144790238950189411257159764069717757857428790821097775857711009224551562331996527320648038253975136644387140098790259506666099526285371437465848910819490249153285257412062781915544119953146199033917714574746100987017044039929593258100845635600270118512899958122332882448857057796244584373161157027836164745303528357442885839453983566556065627037664475925390777502885068243652712586849629195814088165409888631732333157319461798948552717235442541183799269069034647640189514753977972211978021904916322497259494272419187383310422594344365317754635238021011312562417303801,65537}},#{<<"alg">>=><<"RSA-OAEP">>,<<"kid">>=><<"8Djc_h2SS8fTlrn1eXq2FfxCIxLAckCo6vMmJEBn6r4">>,<<"use">>=><<"enc">>,<<"x5c">>=>[<<"MIIClzCCAX8CBgGK+AhcNTANBgkqhkiG9w0BAQsFADAPMQ0wCwYDVQQDDARkZXNwMB4XDTIzMTAwNDAwMTE1MloXDTMzMTAwNDAwMTMzMlowDzENMAsGA1UEAwwEZGVzcDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAK4CDPlTdnIQRnkUwr0B5aC08Xo7zUDfBoDN9JHcuTDfO7xULhkASPZIFzxsJIKnqrWCdQfuWSdAEZm8KPbN43RzpOVADdZECam2n6mG+br8N/vJKfVpOAW/KoIyiK85v/ldMmJJaeiafTxpbLXHkkmRtvWj+BfkhDSj9iJnfuTp+PA7+Cv9NDWltYnVScrjr2Wxaa1aVY1VVpGGHrAUTGBzcdRnbGqOFt9OZGQJXF16DUx5ItjhxPkvL7EWrcbjLoy1sb/VNQX4QAVMaBcFYQKsbax3HraoZW7cvUdXa51pnktGqQzcq77mJq4uDMmvb9QYDatlJN+xmvMzDc9WAPkCAwEAATANBgkqhkiG9w0BAQsFAAOCAQEAFNGbKNoCqRtH+t06JqCHDbN6HwSKG1YXvf5I2F/sw5T0JJpOi8wU7ks+ifDa64MhBFGwVJh4ef65NsWwlRQ7xyAqzXOyoS/nMosHUW+BDQNQ+xgy9PbHPtY/7naTmWoQF1tOBnhk5+wSHgNXCPqUCeQHnAiFAGufZWxU3eq5Q/HPUX39ORyRaoq5G5wyh9khI2tWG2rLKM5KnJSW9aEs+0wliCVF9XHXVHPv9aGLBw7po6TUXpq2vFq4RN8WEyHvNg5ZKbSG078RlODsN55tWTaZE7kA6shb0shwcfJMF0wCPQfbuEIHnv04h1TRvA9F14lUhY2+r6QdOS+IWODfTQ==">>],<<"x5t">>=><<"7_xGCLqWF5M3eHy9uWbSvh5KpXE">>,<<"x5t#S256">>=><<"wUIXStAh3BtRZXx0tEQHK5N8OmqQQwVJGfbMiEtENVE">>}}
```

The logic goes way past my head, but I see more than one key when I call `Oidcc.ProviderConfiguration.Worker.get_jwks`, but it seems to me that this fails validation when the first one doesn't match? I guess it should be testing the other keys in the JWK set.

**Reply by 1player** [reply]
For clarity:

```
iex(8)> Oidcc.ProviderConfiguration.Worker.get_jwks(Gateway.OidccConfigProvider)
#JOSE.JWK<
  keys: {:jose_jwk_set,
   [
     {:jose_jwk, :undefined,
      {:jose_jwk_kty_rsa,
       {:RSAPublicKey,
        21966476291801841635828048148208692901386442529292144790238950189411257159764069717757857428790821097775857711009224551562331996527320648038253975136644387140098790259506666099526285371437465848910819490249153285257412062781915544119953146199033917714574746100987017044039929593258100845635600270118512899958122332882448857057796244584373161157027836164745303528357442885839453983566556065627037664475925390777502885068243652712586849629195814088165409888631732333157319461798948552717235442541183799269069034647640189514753977972211978021904916322497259494272419187383310422594344365317754635238021011312562417303801,
        65537}},
      %{
        "alg" => "RSA-OAEP",
        "kid" => "8Djc_h2SS8fTlrn1eXq2FfxCIxLAckCo6vMmJEBn6r4",
        "use" => "enc",
        "x5c" => ["MIIClzCCAX8CBgGK+AhcNTANBgkqhkiG9w0BAQsFADAPMQ0wCwYDVQQDDARkZXNwMB4XDTIzMTAwNDAwMTE1MloXDTMzMTAwNDAwMTMzMlowDzENMAsGA1UEAwwEZGVzcDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAK4CDPlTdnIQRnkUwr0B5aC08Xo7zUDfBoDN9JHcuTDfO7xULhkASPZIFzxsJIKnqrWCdQfuWSdAEZm8KPbN43RzpOVADdZECam2n6mG+br8N/vJKfVpOAW/KoIyiK85v/ldMmJJaeiafTxpbLXHkkmRtvWj+BfkhDSj9iJnfuTp+PA7+Cv9NDWltYnVScrjr2Wxaa1aVY1VVpGGHrAUTGBzcdRnbGqOFt9OZGQJXF16DUx5ItjhxPkvL7EWrcbjLoy1sb/VNQX4QAVMaBcFYQKsbax3HraoZW7cvUdXa51pnktGqQzcq77mJq4uDMmvb9QYDatlJN+xmvMzDc9WAPkCAwEAATANBgkqhkiG9w0BAQsFAAOCAQEAFNGbKNoCqRtH+t06JqCHDbN6HwSKG1YXvf5I2F/sw5T0JJpOi8wU7ks+ifDa64MhBFGwVJh4ef65NsWwlRQ7xyAqzXOyoS/nMosHUW+BDQNQ+xgy9PbHPtY/7naTmWoQF1tOBnhk5+wSHgNXCPqUCeQHnAiFAGufZWxU3eq5Q/HPUX39ORyRaoq5G5wyh9khI2tWG2rLKM5KnJSW9aEs+0wliCVF9XHXVHPv9aGLBw7po6TUXpq2vFq4RN8WEyHvNg5ZKbSG078RlODsN55tWTaZE7kA6shb0shwcfJMF0wCPQfbuEIHnv04h1TRvA9F14lUhY2+r6QdOS+IWODfTQ=="],
        "x5t" => "7_xGCLqWF5M3eHy9uWbSvh5KpXE",
        "x5t#S256" => "wUIXStAh3BtRZXx0tEQHK5N8OmqQQwVJGfbMiEtENVE"
      }},
     {:jose_jwk, :undefined,
      {:jose_jwk_kty_rsa,
       {:RSAPublicKey,
        20641313488400331482842333632801827120648023324980096389530700862701360370404698841120176319643573019642405474882155080950523770243170031464586789378669393062578314770025577914201441938986977185512524578849607951026074951871844331827627649161547051998739069901154881631634685986139222295974586734188479166819609201344408111630605028307607188002008845872183238407878748577813547198178160173981203179708487463624778672002995658382898517933623963456306828091884566275405131199208276346376362941670986823812099004824666570814203950013953978453313230931857560487151068013086959930379338878323343402713364686952180466273641,
        65537}},
      %{
        "alg" => "RS256",
        "kid" => "zgeOJ-OVB1Ewd6M5nSDjftW7uXjJkxq17dWaPo7z420",
        "use" => "sig",
        "x5c" => ["MIIClzCCAX8CBgGK+AhbMDANBgkqhkiG9w0BAQsFADAPMQ0wCwYDVQQDDARkZXNwMB4XDTIzMTAwNDAwMTE1MVoXDTMzMTAwNDAwMTMzMVowDzENMAsGA1UEAwwEZGVzcDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAKOCvSvPNgv57kTHhbOM3O1WcWRSY9DiwUEaiC9wJGx/K1J1G9yiQBUENtVgAbH5Ek2Y5vqfDeuFBoRa8xNruu9xMCs2eserfRveP3S3T/vlfO9yWJibCptRTzM1zZv4pwiysKLO2kNoUleavwkF3iixoOJMVekrSSaLoOldVbclf0hiBFjzMM7VsPjos2+UrW7BIpkkffvQ2Hm1xa2EHKtUcGul/eKUNJlGEkI4Jps7Nx9zY7HyzREb+FhoFeP7FLJrF5pCApavf4cH2SFV2OteeilWZYa4UNnm9ufmHop/jHJxHdZJj5eLr/2E9/EPX4n7VagLq+S7oEoTJc4vTWkCAwEAATANBgkqhkiG9w0BAQsFAAOCAQEAAILYI+x7odP428qfgq4h9h1IrGe+lx4HjslZS2EAORbttaymiFxueT5s5Ljgw4e4fwgNADIpNzjOkoOL1S9GQ9VyQoWhZ24ypl2+Z/4Fqav8+Gp5FHuxV06Xy64tE5j4Sz0ocUR7Dk9nJzv9tbz/S2x+5d5KJIGr65gUe6CcOU2GDu15BGm/8fm9cqlvgTMue4tiIObxHgdHHG34Vi5iNUYxbxES0/cJfCySyatTJc9kdbVnzZLQET2GRTAxRvtZ4VVN7ExQabcoKz/+OZgtt2oIin/falkDFFGeaALr9maWKEa9IBoqrGSBx1HALdteVCLXQSExQw6/kYkFZLFUSw=="],
        "x5t" => "91MQNZV0Z1N-6TyBShUyUl1rrzo",
        "x5t#S256" => "Cr5db1sZskJFmCiQdcwsgU4CFZ5v6ZY5ueF2FfvEv10"
      }}
   ]},
  fields: %{},
  ...
```

validate_signature was called with the first kid (8Djc_h2SS8fTlrn1eXq2FfxCIxLAckCo6vMmJEBn6r4) in the set, and immediately failed, even though the second kid (zgeOJ-OVB1Ewd6M5nSDjftW7uXjJkxq17dWaPo7z420) is the right one.

**Reply by maennchen** [reply]
@1player The logic look a bit wonky since it has to go through multiple scenarios:

* The token might have a `kid`, but also might not.
* The JWKs might have a `kid` each, but also might not.
* There's no API in `jose` to check if a key has a chance to work with a given token.

The currently implemented solution should proceed through all keys, but seems not to or ignore the result.

Can you put more debug statements into the error handling and the other case conditions?

https://github.com/erlef/oidcc/blob/ef11cc3187e46503034d4518fe84ec0df9c4495d/src/oidcc_jwt_util.erl#L105

https://github.com/erlef/oidcc/blob/ef11cc3187e46503034d4518fe84ec0df9c4495d/src/oidcc_jwt_util.erl#L110

https://github.com/erlef/oidcc/blob/ef11cc3187e46503034d4518fe84ec0df9c4495d/src/oidcc_jwt_util.erl#L113

**Reply by 1player** [reply]
OK, I have added debugging statements following your instructions.

To recap, the token is signed with kid `zgeOJ-OVB1Ewd6M5nSDjftW7uXjJkxq17dWaPo7z420`, and the JWK set contains two keys, in the following order: `8Djc_h2SS8fTlrn1eXq2FfxCIxLAckCo6vMmJEBn6r4` and `zgeOJ-OVB1Ewd6M5nSDjftW7uXjJkxq17dWaPo7z420`.

The first comparison with key `8Djc_etc` in validate_signature/3 fails and returns `{error, {no_matching_key_with_kid, <<"zgeOJ-OVB1Ewd6M5nSDjftW7uXjJkxq17dWaPo7z420">>}}`, then the second key is tested, and this fails on line 105 with `{error, no_matching_key}`.

Which means that `jose_jwt:verify_strict(Jwks, AllowAlgorithms, Token)` is returning the following value: 

```
  {false, 
   {jose_jwt,#{<<"auth_time">>=>1732622979,<<"azp">>=><<"s3gateway-public">>,<<"exp">>=>1732623279,<<"iat">>=>1732622979,<<"iss">>=><<"redacted">>,<<"jti">>=><<"a8efe2b9-9053-4a97-893d-fb93861a9672">>,<<"preferred_username">>=><<"redacted">>,<<"scope">>=><<"openid">>,<<"session_state">>=><<"eb14150b-a18d-4490-bc68-dbe98c2e6952">>,<<"sid">>=><<"eb14150b-a18d-4490-bc68-dbe98c2e6952">>,<<"sub">>=><<"9dac5c18-afde-4ebb-a7ed-3c3fafdfdab3">>,<<"typ">>=><<"Bearer">>}},
 #{<<"alg">>=><<"RS256">>,<<"kid">>=><<"zgeOJ-OVB1Ewd6M5nSDjftW7uXjJkxq17dWaPo7z420">>,<<"typ">>=><<"JWT">>}
 }
```

Any insight on why this verify_strict call is failing?

**Reply by maennchen** [reply]
@1player Thanks for the details. Do you know what is set in the `AllowAlgorithms`?

**Reply by 1player** [reply]
Well this is embarassing. It is empty. After numerous tests, I forgot to set the `signing_algs` option to validate_jwt in my test case. Now that it is set as shown in the documentation, that validate_signature call works as expected.

Though `validate_jwt` still fails because the access token I'm given doesn't have any `aud` claim (https://github.com/erlef/oidcc/blob/08db6c6826df88b1335bacd216f8d0e33b863019/src/oidcc_token.erl#L1013) and I'm trying to figure out why Keycloak doesn't include that claim in the access token without wasting any more of your time.

Thanks for the patience.

**Reply by maennchen** [reply, marked as answer]
@1player Awesome that you found it! I’m still happy that you reported it. The errors you were getting in this case are wrong.

I’ll open an issue to improve that.

**Reply by Nezteb** [reply]
FYI I ran into a similar issue where I forgot to set `signing_algs`, and the `:no_matching_key_with_kid` error led me astray for a while. The fix ended up just being:

```
opts = %{
  signing_algs: client_context.provider_configuration.token_endpoint_auth_signing_alg_values_supported
}
```

The docs obviously show this, but is there a reason `signing_algs` need to be explicitly passed as options when the `client_context` should already have that information? I assumed incorrectly that the `signing_algs` would otherwise be pulled automatically from `client_context.provider_configuration.token_endpoint_auth_signing_alg_values_supported` if it exists.

@maennchen I created an issue to track this: https://github.com/erlef/oidcc/issues/421

😄

**Reply by maennchen** [reply]
@Nezteb Validating access tokens is outside of the OpenID specification and might have different requirements compared to id token validation (inside specification).

If you want to validate an access token with the exact same rules as an id token, just use the function validating id tokens 😊
